src/app/parcels/batch/BatchDataRow.ts

```diff
@@ -0,0 +1,384 @@
+            feminineProducts: createBooleanGroupFromStrings(data.feminine_products),
+            babyProducts: data.baby_food,
+            nappySize: getNappySize(data.extra_information),
+            petFood: createBooleanGroupFromStrings(data.pet_food),
+            otherItems: createBooleanGroupFromStrings(data.other_items),
+            deliveryInstructions: data.delivery_instructions,
+            extraInformation: parseExtraInfo(data.extra_information),
+            attentionFlag: data.flagged_for_attention ?? false,
+            signpostingCall: data.signposting_call_required ?? false,
+            notes: data.notes,
+        };
+    } else {
+        return null;
+    }
+};
+
+export const reducer = async (
+    state: BatchTableDataState,
+    action: BatchActionType
+): Promise<BatchTableDataState> => {
+    switch (action.type) {
+        case "update_cell": {
+            if (action.payload && action.payload.newRow) {
+                const { rowId, newRow } = action.payload;
+                const updatedRows: BatchDataRow[] = state.batchDataRows.map((row) => {
+                    return row.id === rowId ? newRow : row;
+                });
+                return {
+                    ...state,
+                    batchDataRows: updatedRows,
+                };
+            } else {
+                return state;
+            }
+        }
+        case "add_row": {
+            return state.batchDataRows.length < 99
+                ? {
+                      ...state,
+                      batchDataRows: [
+                          ...state.batchDataRows,
+                          {
+                              id:
+                                  Math.max(
+                                      ...state.batchDataRows.map((row) => {
+                                          return row.id;
+                                      })
+                                  ) + 1,
+                              clientId: null,
+                              data: null,
+                          },
+                      ],
+                  }
+                : state;
+        }
+        case "delete_row": {
+            return action.payload?.rowId !== 0
+                ? {
+                      ...state,
+                      batchDataRows: state.batchDataRows.filter(
+                          (row) => row.id !== action.payload?.rowId
+                      ),
+                  }
+                : state;
+        }
+        case "override_column": {
+            if (action.payload?.newOverrideRow) {
+                const overriddenClientFields = getOverridenFields(
+                    action.payload?.newOverrideRow.data.client
+                );
+                const overriddenParcelFields = getOverridenFields(
+                    action.payload?.newOverrideRow.data.parcel
+                );
+                const { newOverrideRow } = action.payload;
```

Comment:
This feels a bit logically tangled at the moment, hopefully making the types more precise should clarify it somewhat